File: src/main/java/com/github/svegon/capi/event/interact/ReachDistanceCallback.java
package com.github.svegon.capi.event.interact;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.network.ClientPlayerInteractionManager;

@FunctionalInterface
public interface ReachDistanceCallback {
    Event<ReachDistanceCallback> EVENT = EventFactory.createArrayBacked(ReachDistanceCallback.class,
            (interactionManager, unmodified) -> unmodified, listeners -> (interactionManager, unmodified) -> {
        for (ReachDistanceCallback listener : listeners) {
            float ret = listener.getReachDistance(interactionManager, unmodified);

            if (ret != unmodified) {
                return ret;
            }
        }

        return unmodified;
    });

    float getReachDistance(ClientPlayerInteractionManager interactionManager, float unmodified);
}

File: src/client/kotlin/io/github/svegon/mclientapi/client/event/interaction/ReachDistanceCallback.kt
package io.github.svegon.mclientapi.client.event.interaction

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.network.ClientPlayerInteractionManager

fun interface ReachDistanceCallback {
    fun getReachDistance(interactionManager: ClientPlayerInteractionManager?, unmodified: Float): Float

    companion object {
        @JvmField
        val EVENT: Event<ReachDistanceCallback> = EventFactory.createArrayBacked<ReachDistanceCallback>(
            ReachDistanceCallback::class.java,
            ReachDistanceCallback { interactionManager: ClientPlayerInteractionManager?, unmodified: Float -> unmodified }) { listeners: Array<ReachDistanceCallback> ->
            ReachDistanceCallback { interactionManager: ClientPlayerInteractionManager?, unmodified: Float ->
                for (listener in listeners) {
                    val ret = listener.getReachDistance(interactionManager, unmodified)

                    if (ret != unmodified) {
                        return@ReachDistanceCallback ret
                    }
                }
                unmodified
            }
        }
    }
}
